File: devvit-app/src/server/core/urlExtract.ts
// Extract URLs from comment/post bodies + strip Reddit markdown artifacts.
//
// Mirrors `backend/src/canonical/url_extract.rs`. The Rust port uses the
// `linkify` crate; here we use a regex sufficient for the Reddit-comment
// shapes the bot actually sees (anything more elaborate gets bounced through
// the backend's resolver anyway — this is just to decide which URLs from the
// comment to send for canonical-finding).
//
// If parity testing surfaces a real divergence with the Rust impl, swap this
// for the `linkify-it` npm package which closely tracks the Rust behavior.

// Trailing characters stripped from extracted URLs. Same set as the Rust
// `TRAILING_MARKDOWN_CHARS` slice; U+201D is the typographic right-double-
// quote that Reddit sometimes renders for smart-quoted pastes.
const TRAILING_MARKDOWN_CHARS = new Set([
  '?',
  '(',
  ')',
  '[',
  ']',
  '\\',
  ',',
  '.',
  '"',
  '”',
  '`',
  '^',
  '*',
  '|',
  '>',
  '<',
  '{',
  '}',
  '~',
  ':',
  ';',
]);

// Match http(s) URLs up to the next whitespace or angle-bracket. Trailing
// punctuation that we want to strip (`,`, `)`, `.`, etc.) lands inside the
// match and is removed by `removeMarkdown` below — matching what the Rust
// `linkify::LinkFinder` + `remove_markdown` pipeline produces.
const URL_RE = /https?:\/\/[^\s<>"]+/gi;

// Strip trailing markdown punctuation from a URL. Ports
// `praw-python-archive/helpers/utils.py:remove_markdown`.
export function removeMarkdown(url: string): string {
  let end = url.length;
  while (end > 0) {
    const ch = url.charAt(end - 1);
    if (!TRAILING_MARKDOWN_CHARS.has(ch)) break;
    end -= 1;
  }
  return url.slice(0, end);
}

// Extract all unique URLs from a chunk of text, in source order. Duplicates
// are dropped (first occurrence wins) — mirrors `URLExtract(only_unique=True)`
// in the legacy Python and the `HashSet`-dedupe pass in the Rust port.
export function extractUrls(body: string): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  for (const match of body.matchAll(URL_RE)) {
    const cleaned = removeMarkdown(match[0]);
    if (cleaned.length === 0) continue;
    if (seen.has(cleaned)) continue;
    seen.add(cleaned);
    out.push(cleaned);
  }
  return out;
}

File: backend/src/canonical/url_extract.rs
//! Extract URLs from comment/post bodies + strip Reddit markdown artifacts.
//!
//! Ports `praw-python-archive/helpers/utils.py:get_urls` (which used the `urlextract`
//! Python crate) and `remove_markdown`. For us, [`linkify`] does the URL
//! extraction — well-maintained Rust crate, drop-in functional replacement.

use std::collections::HashSet;

use linkify::{LinkFinder, LinkKind};

/// Trailing characters that get stripped from extracted URLs.
///
/// These appear when a URL is embedded in Reddit markdown — `[text](url)`
/// leaves a `)` at the end, `(see https://example.eu.)` leaves a `.`,
/// quote-wrapping leaves `"`, etc. The order doesn't matter; the Python
/// version is a tuple `praw-python-archive/helpers/utils.py:70-71`.
const TRAILING_MARKDOWN_CHARS: &[char] = &[
    '?', '(', ')', '[', ']', '\\', ',', '.', '"', '\u{201D}', // U+201D right-double-quote
    '`', '^', '*', '|', '>', '<', '{', '}', '~', ':', ';',
];

/// Extract all unique URLs from a chunk of text, in source order.
///
/// Each URL has trailing markdown punctuation stripped via [`remove_markdown`].
/// Duplicates are dropped (first occurrence wins). Mirrors the
/// `URLExtract(only_unique=True)` behavior of the legacy Python.
pub fn extract_urls(body: &str) -> Vec<String> {
    let mut finder = LinkFinder::new();
    finder.kinds(&[LinkKind::Url]);

    let mut seen = HashSet::new();
    finder
        .links(body)
        .map(|link| remove_markdown(link.as_str()))
        .filter(|url| !url.is_empty())
        .filter(|url| seen.insert(url.clone()))
        .collect()
}

/// Strip trailing markdown punctuation from a URL.
///
/// Ports `praw-python-archive/helpers/utils.py:remove_markdown`. The Python version
/// loops on `url.endswith(markdown_chars)`; we just pop trailing chars
/// while they're in the trailing-chars set.
pub fn remove_markdown(url: &str) -> String {
    let mut s = url.to_string();
    while let Some(last) = s.chars().last() {
        if TRAILING_MARKDOWN_CHARS.contains(&last) {
            s.pop();
        } else {
            break;
        }
    }
    s
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn extracts_single_url() {
        let urls = extract_urls("check this out: https://example.eu/article");
        assert_eq!(urls, vec!["https://example.eu/article"]);
    }

    #[test]
    fn deduplicates_repeated_urls() {
        let urls = extract_urls("https://example.eu and https://example.eu again");
        assert_eq!(urls, vec!["https://example.eu"]);
    }

    #[test]
    fn preserves_source_order_when_extracting_multiple() {
        let body = "first https://a.example then https://b.example, finally https://c.example";
        let urls = extract_urls(body);
        assert_eq!(
            urls,
            vec![
                "https://a.example",
                "https://b.example",
                "https://c.example"
            ]
        );
    }

    #[test]
    fn strips_trailing_markdown_punctuation() {
        assert_eq!(remove_markdown("https://example.eu."), "https://example.eu");
        assert_eq!(remove_markdown("https://example.eu,"), "https://example.eu");
        assert_eq!(remove_markdown("https://example.eu)"), "https://example.eu");
        assert_eq!(
            remove_markdown("https://example.eu?)"),
            "https://example.eu"
        );
        assert_eq!(
            remove_markdown("https://example.eu\u{201D}"),
            "https://example.eu"
        );
    }

    #[test]
    fn preserves_url_without_trailing_punctuation() {
        assert_eq!(remove_markdown("https://example.eu"), "https://example.eu");
        assert_eq!(
            remove_markdown("https://example.eu/article"),
            "https://example.eu/article"
        );
    }

    #[test]
    fn extracts_from_reddit_link_markdown() {
        // `[text](url)` syntax. linkify captures the URL inside the parens,
        // and remove_markdown strips trailing `)` + `,`.
        let body =
            "see [the article](https://example.eu/news), or directly: https://example.eu/raw";
        let urls = extract_urls(body);
        assert!(urls.contains(&"https://example.eu/news".to_string()));
        assert!(urls.contains(&"https://example.eu/raw".to_string()));
    }

    #[test]
    fn handles_text_without_urls() {
        assert!(extract_urls("no urls here just text").is_empty());
    }

    #[test]
    fn handles_empty_input() {
        assert!(extract_urls("").is_empty());
    }

    #[test]
    fn extracts_unencoded_amp_url_pasted_into_query_string() {
        // The kind of URL that's pasted directly into the API query string
        // (per v7 plan, unencoded URLs work as long as `q` is the last param).
        let body = "https://www.google.com/amp/s/electrek.co/2018/06/19/tesla-model-3/amp/";
        let urls = extract_urls(body);
        assert_eq!(urls.len(), 1);
        assert!(urls[0].contains("/amp/s/electrek.co"));
    }
}
